### ISSUE TITLE
geom_dotplot dot layout with groups

### ISSUE THREAD

**Issue by ironv** [tidyverse/ggplot2#1745]
geom_dotplot is not laying out the points correctly when there are 2 groups.  See http://stackoverflow.com/questions/39375983/geom-dotplot-dot-layout-with-groups

```
require(data.table)
require(ggplot2)

set.seed(10000)
n <- 300

dt <- data.table(Duration=sample(100:800,n,replace=T), EndType=round(runif(n,min=.4)), Group=sample(c("GrpA","GrpB"),n,replace=T))
dt <- rbind(dt, dt[, -c("Group"), with=F][, Group:="All"])
dt[, ":="(Group=factor(Group, levels=c("All","GrpA","GrpB"), ordered=T), EndType=factor(EndType, levels=c(0,1)))]

g <- ggplot(dt, aes_string(x="Group", y="Duration")) + coord_flip() +
    geom_boxplot(aes(ymin=..lower.., ymax=..upper..), fatten=1.1, lwd=.1, outlier.shape=NA) + 
    geom_dotplot(aes(fill=EndType), binaxis="y", stackdir="center", stackgroups=T, method="histodot", binwidth=15, dotsize=.5) +
    scale_fill_manual(values=c("white","black"))
print(g)
```

**Comment by hadley** [comment]
Can you please create a minimal reprex?

**Comment by Ax3man** [comment]
A more minimal reprex:
```
ggplot(mtcars, aes(x = factor(gear), y = mpg, fill = factor(cyl))) +
  geom_dotplot(binaxis = "y", stackdir = "center",  stackgroups = TRUE,  method = "histodot")
```
It seems to me that when `stackgroups = TRUE` the stacking is done for all points together, without separation by the grouping on the x-axis. This can cause dots to 'float away' from the center.

When using `stackdir = 'up'` it is fine for the first cloud, but gets progressively worse towards the right.
```
ggplot(mtcars, aes(x = factor(gear), y = mpg, fill = factor(cyl))) +
  geom_dotplot(binaxis = "y", stackdir = "up", stackgroups = TRUE,  method = "histodot")
```

**Comment by hadley** [comment]
@wch any ideas?

**Comment by wch** [comment]
For this to work, the dotplot needs to be able do a calculation spanning all x groups, but within other grouping variables. I think only the x variable is special this way.

**Comment by eknud** [comment]
Has a solution been presented here? Unclear if we've solved for this yet.

**Comment by thomasp85** [comment]
Can users of `geom_dotplot()` please stress test the fix in #4417 - While it fixes the issue reported here I am unsure whether it has regressions somewhere else
